railties/lib/rails/application/routes_reloader.rb

```diff
@@ -19,23 +19,11 @@ def reload!
         clear!
         load_paths
         finalize!
+        route_sets.each(&:eager_load!) if eager_load
```

Comment:
Even that this seems cheap I want to avoid these calls on every routes reload. I think a alternative implementation would be remove the `updater.execute` from inside the `updater` method. Can you see if that fix it?

Reply:
Currently, `updater.execute` is executed on every app boot to load the routes. I believe this is necessary, if I remove it, the routes are not finalized and thus don't work at all. The routes need to be eager loaded here too.

Reply:
Any thoughts @rafaelfranca ?

Reply:
Anything I can do to move this along? It fixes my app app's routes getting corrupted in production.

Reply:
That is strange. We are relying on the `updater` to call `updater.execute` in the first time it is called, but we also call `execute` on the router reload on boot time, so why calling `updater.execute` explicitly on the `execute` method don't work?

Reply:
Can you provide a test case for this? Or at least an application were the absent of this code will break something? I want to try myself to understand what is happening here.

Reply:
The problem is that the original `execute_if_updated` method was added and called from [here](https://github.com/rails/rails/blob/master/railties/lib/rails/application/finisher.rb#L130) with the assumption that it would execute at application boot and eager load the routes. What happens is `updater.execute_if_updated` returns `false` so the routes are not eager loaded. I will put together an example app. I'm not sure where to start writing tests for this.

Reply:
Also, the `execute` & `execute_if_updated` methods are redundant, all they do is eager load if the updater updated. The updater just runs `reload!` so eager loading should just take place in `reload!`.

Reply:
Here's the [test app](https://github.com/jboler/routes_reloader_test). The second commit shows the changes from a plain `rails new` app.

Reply:
This patch seems to fix the issue you are trying to.

```diff
diff --git a/railties/lib/rails/application/finisher.rb b/railties/lib/rails/application/finisher.rb
index c4b188aeee..04aaf6dd9a 100644
--- a/railties/lib/rails/application/finisher.rb
+++ b/railties/lib/rails/application/finisher.rb
@@ -127,7 +127,7 @@ def self.complete(_state)
       initializer :set_routes_reloader_hook do |app|
         reloader = routes_reloader
         reloader.eager_load = app.config.eager_load
-        reloader.execute_if_updated
+        reloader.execute
         reloaders << reloader
         app.reloader.to_run do
           # We configure #execute rather than #execute_if_updated because if
diff --git a/railties/lib/rails/application/routes_reloader.rb b/railties/lib/rails/application/routes_reloader.rb
index 2ef09b4162..6c7a8c732e 100644
--- a/railties/lib/rails/application/routes_reloader.rb
+++ b/railties/lib/rails/application/routes_reloader.rb
@@ -41,7 +41,6 @@ def execute_if_updated
       def updater
         @updater ||= begin
           updater = ActiveSupport::FileUpdateChecker.new(paths) { reload! }
-          updater.execute
           updater
         end
       end
```

Reply:
Yes, that was one of my earlier solutions and I refactored it to what it is now because the `execute` & `execute_if_updated` methods don't need to exist and in my opinion make RoutesReloader more unreadable.

Eager loading the routes in `reload!` is the same end result as eager loading in the `execute` & `execute_if_updated` methods because those methods just run `reload!` via the updater.

Reply:
I believe the ideal PR would combine your patch with my PR.

Reply:
Originally the `execute` & `execute_if_updated` methods were delegated to `updater` but I believe it was erroneously changed [here](https://github.com/rails/rails/pull/27647/files#diff-24dc69a82eadb00e5a2a8a776b9399eaL7).

Reply:
I believe we also want the route eager loading to be run if `Rails.application.reload_routes!` is run for any reason. If we only trigger the eager loading from the updater then `reload_routes!` won't be consistent:

```
    # Reload application routes regardless if they changed or not.
    def reload_routes!
      routes_reloader.reload!
    end
```
https://github.com/rails/rails/blob/master/railties/lib/rails/application.rb#L167

Reply:
I really think we should not call `eager_load!` every time the routes are reloaded. We only want to call this method once given this method is to make sure everything that should be loaded on boot is being loaded on boot while reload is to make sure everything that was loaded is unloaded and loaded again.

Reply:
oh, I forgot that in development where reload is called multiple times `eager_load` is false. So my concern about `eager_load!` being called multiple times is not true.